Sources/Sentry/SentryViewHierarchy.m

```diff
@@ -1,48 +1,140 @@
+#    define tryJson(code)                                                                          \
+        if ((result = (code)) != SentryCrashJSON_OK)                                               \
+        goto done
+
+- (BOOL)processViewHierarchy:(NSArray<UIView *> *)windows
+                 addFunction:(SentryCrashJSONAddDataFunc)addJSONDataFunc
+                    userData:(void *const)userData
+{
+
+    SentryCrashJSONEncodeContext JSONContext;
+    sentrycrashjson_beginEncode(&JSONContext, false, addJSONDataFunc, userData);
+
+    int result;
+
+    tryJson(sentrycrashjson_beginObject(&JSONContext, NULL));
+    tryJson(sentrycrashjson_addStringElement(
+        &JSONContext, "rendering_system", "UIKIT", SentryCrashJSON_SIZE_AUTOMATIC));
+    tryJson(sentrycrashjson_beginArray(&JSONContext, "windows"));
+
+    for (UIView *window in windows) {
+        tryJson([self viewHierarchyFromView:window intoContext:&JSONContext]);
+    }
+
+    tryJson(sentrycrashjson_endContainer(&JSONContext));
+
+    result = sentrycrashjson_endEncode(&JSONContext);
+
+done:
+    if (result != SentryCrashJSON_OK) {
+        SENTRY_LOG_DEBUG(
+            @"Could not create view hierarchy json: %s", sentrycrashjson_stringForError(result));
+        return false;
+    }
+    return true;
+}
+
+- (int)viewHierarchyFromView:(UIView *)view intoContext:(SentryCrashJSONEncodeContext *)context
 {
-    [[self fetchViewHierarchyPreventMoveToMainThread:YES]
-        enumerateObjectsUsingBlock:^(NSString *description, NSUInteger idx, BOOL *stop) {
-            NSString *fileName =
-                [NSString stringWithFormat:@"view-hierarchy-%lu.txt", (unsigned long)idx];
-            NSString *filePath = [path stringByAppendingPathComponent:fileName];
-            NSData *data = [description dataUsingEncoding:NSUTF8StringEncoding];
-            [data writeToFile:filePath atomically:YES];
-        }];
+    int result = 0;
+    tryJson(sentrycrashjson_beginObject(context, NULL));
+    const char *viewClassName = [[SwiftDescriptor getObjectClassName:view] UTF8String];
+    tryJson(sentrycrashjson_addStringElement(
+        context, "type", viewClassName, SentryCrashJSON_SIZE_AUTOMATIC));
+
+    if (view.accessibilityIdentifier && view.accessibilityIdentifier.length != 0) {
+        tryJson(sentrycrashjson_addStringElement(context, "identifier",
+            view.accessibilityIdentifier.UTF8String, SentryCrashJSON_SIZE_AUTOMATIC));
+    }
+
+    tryJson(sentrycrashjson_addFloatingPointElement(context, "width", view.frame.size.width));
+    tryJson(sentrycrashjson_addFloatingPointElement(context, "height", view.frame.size.height));
+    tryJson(sentrycrashjson_addFloatingPointElement(context, "x", view.frame.origin.x));
+    tryJson(sentrycrashjson_addFloatingPointElement(context, "y", view.frame.origin.y));
+    tryJson(sentrycrashjson_addFloatingPointElement(context, "alpha", view.alpha));
+    tryJson(sentrycrashjson_addBooleanElement(context, "visible", !view.hidden));
```

Comment:
This way we can write directly to file, or to memory buffer, no need to first create a dictionary, and then serialize the dictionary.
This will be used during signal handling, the faster and less memory allocation being done the better.

Reply:
Yes, but we already call into objective C code to make this work. We should only call async safe code in a signal handler, but as we are already violating that contract, I'm unsure if it's worth the effort. We call this only after SentryCrash wrote the crash report. In some cases, we will crash when calling into ObjC, but I think it doesn't matter how often we do it. Once is enough.

Reply:
Is not about safety, because this code itself is called from an objc class. Is about performance, this is the fastest way to do it, and since we may deal of view hierarchy with hundreds of views, the faster the better.

Reply:
How much faster? Do you have benchmarks?

Reply:
You convinced me with `during signal handling`.